### PR TITLE
feat: Implement Pervasives.compare

### LINKED ISSUE #305: Check Safety and Efficiency of Numeric Comparison Functions
Currently, our numeric comparison functions (sans `==`) are implemented by coercing numbers into `f64`s and doing the check. This works well enough, but we could probably do it more efficiently if we wrote out a proper (large) implementation of each function. Moreover, we should double-check that we aren't losing anything important precision-wise with this conversion (specifically around the relative ordering of very large `i64` values). **We should be able to guarantee to users that the comparison of two numbers of the same representation (e.g. (i64,i64), (i32, i32), etc.) is correct.**

### PR DESCRIPTION
**Pull request by ospencer**
`compare` can be used for sorting or other kinds of ordering operations. Works on all types. This also improves the accuracy of all of our comparison operators.

Closes #305

### CODE REVIEW

**Review by marcusroberts — APPROVED** [review]
Looks excellent!

File: stdlib/pervasives.gr
Comment (phated) [inline stdlib/pervasives.gr:229]: It's not just `-1` and `1`? Is there some sort of scale?

File: stdlib/runtime/compare.gr
Comment (phated) [inline stdlib/runtime/compare.gr:27]: Is there a reason this import is hanging out alone?

File: stdlib/runtime/numbers.gr
Comment (phated) [inline stdlib/runtime/numbers.gr:1834]: So `16/16 == 1` is never true? I guess it makes sense but also seems a little weird (especially if I'm adding rationals to get to `1`)

File: compiler/test/stdlib/pervasives.test.gr
Comment (phated) [inline compiler/test/stdlib/pervasives.test.gr:6]: I think we should have a test for `NaN` (given the code comments throughout) and maybe even `Infinity` (?)

**Review by phated — COMMENTED** [review]
I *love* polymorphic compare! A few questions and comments before approval

File: compiler/test/stdlib/pervasives.test.gr
Comment (ospencer) [reply compiler/test/stdlib/pervasives.test.gr:16]: I can add some 🙂

File: stdlib/pervasives.gr
Comment (ospencer) [reply stdlib/pervasives.gr:229]: There's no scale! I believe languages implement it this way just to make it exhaustive and allow simple implementations like `a - b` that don't need any branching logic. I believe this implementation will ever only return one of `-1, 0, 1` though (I'd need to double-check a few things).

File: stdlib/runtime/compare.gr
Comment (ospencer) [reply stdlib/runtime/compare.gr:27]: Copypasta from `equal` that did it the same way 😛

File: stdlib/runtime/numbers.gr
Comment (ospencer) [reply stdlib/runtime/numbers.gr:1834]: This might need some clarification! An invariant of the number system is that rationals with a denominator of `1` do not exist, so writing the literal `16/16` gets you `1/1` which Grain just simplifies to simply `1`. So `compare(16/16, 1)` will always be zero, but this case doesn't need to worry about it because of the invariants. They're all listed in this file, but I think it might be nice to have one paragraph at the start that explains them. Shall I make an issue to get that added?

File: compiler/test/stdlib/pervasives.test.gr
Comment (ospencer) [reply compiler/test/stdlib/pervasives.test.gr:6]: I could have sworn I added some!

File: stdlib/runtime/numbers.gr
Comment (phated) [reply stdlib/runtime/numbers.gr:1835]: Sure, an issue would be nice.

File: stdlib/runtime/compare.gr
Comment (phated) [reply stdlib/runtime/compare.gr:27]: Looks like you updated `equal` but not this file 🙃

**Review by phated — APPROVED** [review]
Still needs a rebase but LGTM!

**Pull request merged**
